### PR TITLE
[SIM_FLUTTER-53] [Integrate] Display the lists of transactions

### PR DESCRIPTION
**Pull request by IanUrriza10**
### Issue Links
- [backlog](https://framgiaph.backlog.com/view/SIM_FLUTTER-53)
- [slack](https://sun-philippine.slack.com/archives/C059L7XAXFB/p1688966800410359)

### Definition of done
- [x] Display All types of transactions correctly
    - [x] Correct Receiver for 'Transferred money to'
    - [x] Correct Sender for 'Transferred money from'
    - [x] Correct AccountName for  'Credit and Dept'
- [x] Correct  Date and Time
- [x] Correct Sign for positive and negative values
    - [x] Credit = -
    - [x] Deposit = +
    - [x] Transferred from = +
    - [x] Transferred to = -     
### Screenshots
image
image

### CODE REVIEW

File: frontend/lib/src/controllers/transaction_controller.dart [hunk 1]
```diff
@@ -18,6 +17,15 @@ class TransactionController extends GetxController {
 
   bool get transactionSubmitEnabled => _transactionSubmitEnabled.value;
 
+  RxList<Transaction> transactionList = <Transaction>[].obs;
```
Comment (jstephend-sun) [inline frontend/lib/src/controllers/transaction_controller.dart]: move to top, pa compress din ng variables

_selectedTransactionType
_selectedTransactionDate
_transactionSubmitEnabled
transactionList

without spaces

File: frontend/lib/src/features/transaction_history/transaction_history.dart [hunk 2]
```diff
@@ -13,15 +12,18 @@ import 'package:frontend/src/features/dashboard/components/account_card.dart';
...
+
+  @arguments accountId, contains Id to access account details
 */
 class TransactionHistory extends StatelessWidget {
   const TransactionHistory({super.key});
 
   @override
   Widget build(BuildContext context) {
-    final transactionController = Get.put(TransactionController());
+    final controller = Get.put(TransactionController());
```
Comment (jstephend-sun) [inline frontend/lib/src/features/transaction_history/transaction_history.dart]: add Type on the left hand

**Comment by Dbulawan13** [comment]
Comment:

- [ ] Please remove "+" button in both Admin and User accounts


- [ ] When user/admin deposit money. The text should be: `Deposited money to`


- [ ] When user/admin received money from other users/admin accounts. The text should be: `Received money from`


Comments: 3

File: frontend/lib/src/models/transaction.dart [hunk 3]
```diff
@@ -11,37 +12,66 @@ enum TransactionTypes { CREDIT, DEPT, TRANSFER }
...
+  String? accountName;
+  String? senderName;
+  String? receiverName;
 
   Transaction({
-    this.name,
-    this.date,
-    required this.amount,
-    required this.description,
+    this.transactionId,
```
Comment (jstephend-sun) [inline frontend/lib/src/models/transaction.dart]: use id only, its the model's id anyways

File: frontend/lib/src/controllers/transaction_controller.dart [hunk 4]
```diff
@@ -1,5 +1,4 @@
 import 'package:get/get.dart';
```
Comment (jstephend-sun) [inline frontend/lib/src/controllers/transaction_controller.dart]: space got removed

File: frontend/lib/src/features/transaction_history/components/transaction_card.dart [hunk 5]
```diff
@@ -2,6 +2,7 @@ import 'package:flutter/material.dart';
 import 'package:font_awesome_flutter/font_awesome_flutter.dart';
 
 import 'package:frontend/src/models/transaction.dart';
+import 'package:intl/intl.dart';
```
Comment (jstephend-sun) [inline frontend/lib/src/features/transaction_history/components/transaction_card.dart]: sort import

File: frontend/lib/src/features/transaction_history/components/transaction_card.dart [hunk 6]
```diff
@@ -12,7 +13,6 @@ class TransactionCard extends StatelessWidget {
   const TransactionCard({super.key, required this.transaction});
 
   final Transaction transaction;
```
Comment (jstephend-sun) [inline frontend/lib/src/features/transaction_history/components/transaction_card.dart]: space got removed

File: frontend/lib/src/features/transaction_history/components/transaction_card.dart [hunk 7]
```diff
@@ -26,7 +26,6 @@ class TransactionCard extends StatelessWidget {
             color: Color(0xFF6D7881),
           ),
         );
```
Comment (jstephend-sun) [inline frontend/lib/src/features/transaction_history/components/transaction_card.dart]: space got removed

File: frontend/lib/src/features/transaction_history/transaction_history.dart [hunk 8]
```diff
@@ -13,15 +12,18 @@ import 'package:frontend/src/features/dashboard/components/account_card.dart';
...
     final arguments =
         ModalRoute.of(context)!.settings.arguments as ScreenArguments;
+    controller.getTransactions(accountId: arguments.accountId);
```
Comment (jstephend-sun) [inline frontend/lib/src/features/transaction_history/transaction_history.dart]: implement using FutureBuilder

File: frontend/lib/src/models/transaction.dart [hunk 9]
```diff
@@ -11,37 +12,66 @@ enum TransactionTypes { CREDIT, DEPT, TRANSFER }
...
+        "date": transactionDate,
+        "transaction_type": transactionType,
+        "category": category,
         "amount": amount,
         "description": description,
-        "transaction_type": transactionType,
-        "category": category
+        "accountName": accountName
       };
+  String generateDescription() {
```
Comment (jstephend-sun) [inline frontend/lib/src/models/transaction.dart:60]: space before function

File: frontend/lib/src/services/transaction_service.dart [hunk 10]
```diff
@@ -1,6 +1,5 @@
 import 'dart:convert';
 import 'dart:io';
```
Comment (jstephend-sun) [inline frontend/lib/src/services/transaction_service.dart]: space got removed

File: backend/app/Http/Controllers/TransactionController.php [hunk 11]
```diff
@@ -21,9 +21,13 @@ public function index(FilterTransactionRequest $request, Account $account = null
...
         // Get transactions by all or through accounts
         // For shallow nesting see https://laravel.com/docs/8.x/controllers#shallow-nesting
-        $transactions = $account
-            ? Transaction::where('account_id', $account['id'])->orderBy('transaction_date')->get()
-            : Transaction::orderBy('transaction_date')->get();
+        $transactions = ($account ? Transaction::where('account_id', $account['id']) : Transaction::all())
+            ->with(['account'])
+            ->orderBy('transaction_date')
+            ->get();
+        // $transactions = $account
```
Comment (jstephend-sun) [inline backend/app/Http/Controllers/TransactionController.php]: unnecessary codde

File: frontend/lib/src/models/transaction.dart [hunk 12]
```diff
@@ -11,37 +12,66 @@ enum TransactionTypes { CREDIT, DEPT, TRANSFER }
...
     required this.transactionType,
     required this.category,
+    required this.amount,
+    required this.description,
+    this.accountName,
+    this.senderName,
+    this.receiverName,
   });
 
+  // DateFormat('yyyy-MM-ddTHH:mm:ssZ').parseUTC(json["created_at"]),
```
Comment (jstephend-sun) [inline frontend/lib/src/models/transaction.dart]: unnecessary code

**Review by jstephend-sun — CHANGES_REQUESTED** [review]
Requested Changes #56

review time: 15mins
comments: 10

File: frontend/lib/src/controllers/transaction_controller.dart [hunk 13]
```diff
@@ -7,17 +7,23 @@ List<String> transactionTypes = ['Credit', 'Dept', 'Transfer'];
...
   final RxBool _transactionSubmitEnabled = false.obs;
+  RxList<Transaction> transactionList = <Transaction>[].obs;
 
   String get selectedTransactionType => _selectedTransactionType.value;
 
   DateTime get selectedTransactionDate => _selectedTransactionDate.value;
 
   bool get transactionSubmitEnabled => _transactionSubmitEnabled.value;
 
+  Future<List<Transaction>> getTransactions({String accountId = ''}) async {
```
Comment (jstephend-sun) [inline frontend/lib/src/controllers/transaction_controller.dart]: organize the position of this function, its in the middle differing code sections

File: frontend/lib/src/models/transaction.dart [hunk 14]
```diff
@@ -11,37 +12,66 @@ enum TransactionTypes { CREDIT, DEPT, TRANSFER }
...
+  String? accountName;
+  String? senderName;
+  String? receiverName;
 
   Transaction({
-    this.name,
-    this.date,
-    required this.amount,
-    required this.description,
+    this.id,
```
Comment (jstephend-sun) [inline frontend/lib/src/models/transaction.dart:26]: shouldnt id and transaction date be required here?

**Comment by Dbulawan13** [comment]
Comment:

- [ ] There is no transaction record, it should display transaction records.
image

Database and in Previous commit:
image



Comment: 1

**Review by jstephend-sun — CHANGES_REQUESTED** [review]
Requested Changes #56

check out unresolved comment on my previous review, it wasnt addressed

total review time: 25mins
total comments: 10 + 3
iteration: 2nd review

File: frontend/lib/src/models/transaction.dart (hunk 14 above)
Comment (IanUrriza10) [reply frontend/lib/src/models/transaction.dart:26]: postTransaction doesn't have any id or dates yet

Comment (jstephend-sun) [reply frontend/lib/src/models/transaction.dart:26]: were talking about the model here

**Comment by Dbulawan13** [comment]
Comment:

- [x] When user/admin deposit money. The text should be: `Deposited money to`


- [x] When user/admin transfer money to another user. The text should be: `Transferred money to`


Previous Commit:
image

Comments: 2

**Review by Dbulawan13 — APPROVED** [review]
Approve PR #56 

Working;
- Able to display lists of transactions
- Display correct text
  - **Deposited money to {own_account_name} :** When user Deposit money to his/her account
  - **Deducted money from {own_account_name} :**  When user Withdraw money to his/her account
  - **Transferred money to {another_user_account_name} :** When user transfer an amount
  - **Received money From {another_user_account_name} :** When user received an amount
- Correct date has been displayed
- Transaction history will change in realtime
- Display **No transaction Record** if there is no transaction history





Testing hours: 2
Comments: 6
Iterations: 3

**Review by jstephend-sun — APPROVED** [review]
Approved #56

total review time: 25mins
total comments: 10 + 3
iteration: 2nd review

**Pull request merged**